alf/algorithms/muzero_algorithm.py

```diff
@@ -295,70 +295,182 @@ def _hook(grad, name):
     @torch.no_grad()
     def preprocess_experience(self, root_inputs: TimeStep,
                               rollout_info: MCTSInfo, batch_info):
-        """Fill rollout_info with MuzeroInfo
+        """Fill rollout_info with MuzeroInfo.
+
+        Especially, the training targets for representation learning is computed
+        here with reanalyze and/or bootstrapping.
+
+        Note that the shape of experience is [B, T, ...], where B is the batch
+        size T is the mini batch length.
 
-        Note that the shape of experience is [B, T, ...]
         """
         assert batch_info != ()
         replay_buffer: ReplayBuffer = batch_info.replay_buffer
         info_path: str = "rollout_info"
         info_path += "." + self.path if self.path else ""
-        mini_batch_length = root_inputs.step_type.shape[1]
         rollout_value = rollout_info.value
-        assert mini_batch_length == 1, (
-            "Only support TrainerConfig.mini_batch_length=1, got %s" %
-            mini_batch_length)
-
         value_field = info_path + '.value'
         candidate_actions_field = info_path + '.candidate_actions'
         candidate_action_policy_field = (
             info_path + '.candidate_action_policy')
 
+        # Create aliases for mini_batch_size (B), mini_batch_length(T) and
+        # predictive unroll steps (R) to make the implementation below more
+        # succinct.
+        B, T = root_inputs.step_type.shape
+        R = self._num_unroll_steps
+
+        # The implementation below heavily relies on the concept of "unfold".
+        # Take an individual trajectory of size T = 3 as example. Assuming the
+        # steps in this trajectory corresponds to position 11, 12, 13 in the
+        # replay buffer. In order to compute a certain type of targets for this
+        # trajectory with an unroll steps R = 5, a naive implementation will
+        # need to do computation for 15 T * (R + 1) steps at positions
+        #
+        #    11, 12, 13, 14, 15, 16
+        #    12, 13, 14, 15, 16, 17
+        #    13, 14, 15, 16, 17, 18
+        #
+        # respectively. Here we can employ a trick to compute only T + R + 1
+        # times at positions
+        #
+        #    11, 12, 13, 14, 15, 16, 17, 18
+        #
+        # and transform it to the above form. Such transformation that converts
+        # a shape of [T + R] to [T, R + 1] is the "unfold" operations.
+        #
+        # We also name the [T + R] shaped positions as "folded" positions.
+
+        def _unfold_dim1(x: torch.Tensor) -> torch.Tensor:
+            """Perform the aforementioned unfold at dim = 1 of the input tensor.
+
+            """
```

Comment:
Not sure how fast unfold is. But if T == 1, we can simply `return x.unsqueeze(1)` to save some time.

Reply:
Did not dig into the implementation of `unfold`, but adding two lines to specialize `T == 1` sounds good to me. Added.